### DISCUSSION TITLE
In what situation not having a Node location is actually a thing?

### DISCUSSION THREAD

**Discussion by ericmorand** [eslint/eslint#18811] [General] [answered]
There is one thing that has been bugging me forever: the node passed as parameter to the various listeners of `NodeListener` are not guaranteed to have a location (`loc`). I'm not sure I understand the rational here: in what situation would a rule not want to emit the location where an issue was detected?

The current situation forces everybody to add an impossible to cover `else` branch just to ensure that `loc` is defined:

```ts
const ruleModule: Rule.RuleModule = {
    meta: {},
    create: (context) => {
        return {
            "Program:exit": (node) => {
                if (node.loc) {
                    context.report({
                        message: 'Something is wrong',
                        loc: node.loc
                    });
                }
                else {
                   // what are we supposed to do?
                }
            }
        };
    }
}
```

Of course, it is possible to force type the loc as defined (`node.loc!`) but if we have to do that everytime we develop a rule, why is it allowed to be undefined to begin with?

What is the exact benefit of having the `loc` property allowed to be undefined and what situation does it solve?

**Comment by nzakas** [comment, marked as answer]
The built-in parser guarantees that all nodes have a `loc` property, and indeed, ESLint expects this. If you're using the built-in parser and you find a node doesn't have a `loc`, then please open an bug report.

If you are using a different parser (maybe `typescript-eslint`?), that's also likely a bug. Every node should have a `loc`. I'd suggest filing a bug report with that parser.

**Reply by ericmorand** [reply]
My point is that the nodes are declared as possibly not having a loc. You confirm that they all have one, hence I think the question is even more legit: why is the Node type `loc` property declared as optional if it is guaranteed to be present?

**Reply by nzakas** [reply]
Ah okay, so you're talking about the type definitions and not the functional code.

AFAIK, `@types/eslint` (which we do not maintain) is using types from the `estree` package as the basis for the nodes. ESTree does not guarantee any location information to be present, so those types have `loc` and `range` as optional even though ESLint requires them.

**Reply by ericmorand** [reply]
Make sense. I'll try to contact the maintainer of the types package or propose a PR myself. Thanks a lot for your time. 🙏
